src/clj_irods/core.clj

```diff
@@ -0,0 +1,205 @@
+            listing-item (when (delay? paged-folder-listing) (get-from-listing (apply concat (icat/flatten-cached-listings @paged-folder-listing))))]
+        (or
+          (when listing-item (delay (extract-fn listing-item)))
+          (when single-item  (delay (extract-fn @single-item)))
+          nil))
+      (or
+        ;; better if we pass user group IDs in so we can get them from any source, rather than forcing the icat version specifically
+        (when (:has-icat irods) (delay (force (icat/user-group-ids irods user zone)) (extract-fn @(icat/get-item irods user zone path))))
+        nil))))
+
+(defn- from-stat
+  [cache? irods extract-fn path]
+  (if cache?
+    (let [jargon-stat (jargon/cached-stat irods path)]
+      (or
+        (when jargon-stat (delay (extract-fn @jargon-stat)))
+        nil))
+    (or
+      (when (:has-jargon irods) (delay (extract-fn @(jargon/stat irods path))))
+      nil)))
+
+(defn- from-stat-or-listing
+  [stat-extract-fn listing-extract-fn irods user zone path]
+  (cached-or-get irods
+    [from-listing listing-extract-fn user zone path]
+    [from-stat stat-extract-fn path]))
+
+(defn object-type
+  [irods user zone path]
+  (let [listing-extract-fn (fn [item] (condp = (:type item) "dataobject" :file "collection" :dir nil))]
+    (from-stat-or-listing :type listing-extract-fn irods user zone path)))
+
+(defn date-modified
+  [irods user zone path]
+  (from-stat-or-listing :date-modified (comp (partial * 1000) #(Integer/parseInt %) :modify_ts) irods user zone path))
+
+(defn date-created
+  [irods user zone path]
+  (from-stat-or-listing :date-created (comp (partial * 1000) #(Integer/parseInt %) :create_ts) irods user zone path))
+
+(defn file-size
+  [irods user zone path]
+  (from-stat-or-listing :file-size :data_size irods user zone path))
+
+(defn uuid
+  [irods user zone path]
+  (cached-or-get irods
+    [from-listing :uuid user zone path]
+    [(fn [cache? irods path]
+       (let [get-from-metadata (fn [metadata]
+                                 (when-let [uuid-meta (first (filter #(= (:attr %) "ipc_UUID") @metadata))]
+                                   (delay (get uuid-meta :value))))]
+           (when-let [metadata (if cache?
+                                 (jargon/cached-get-metadata irods path)
+                                 (jargon/get-metadata irods path :known-type @(object-type irods user zone path)))]
+             (or
+               (get-from-metadata metadata)
+               nil))))
+     path]))
+
+(defn permission
+  [irods user zone path]
+  (cached-or-get irods
+    [from-listing (comp jargon-perms/fmt-perm :access_type_id) user zone path]
+    [(fn [cache? irods user path]
+       (if cache?
+         (jargon/cached-permission-for irods user path)
+         (delay
+           @(jargon/permission-for irods user path :known-type @(object-type irods user zone path)))))
+     user path]))
+
+(defn folder-listing
```

Comment:
This one's a doozy. Most of the subparts I'll explain in the `clj-irods.icat` namespace, but the main summary is:

* fetch all cached listings that match the entity type, info-types, sort-column, and sort-direction of this request.
* when there's several whose limits/offsets work out to adjacent ranges (e.g. limit 10 offset 0 and limit 10 offset 10), combine them into one (e.g. limit 20 offset 0)
* if the cached, combined values can satisfy the limit/offset being requested, return those from the cache
* otherwise, do a new request

There's two potential improvements here:

* rather than requiring specifically abutting ranges, also merge overlapping ranges. I didn't do this because overlapping ranges are probably a pretty rare case (even multiple listings in one `with-irods` block is probably pretty rare)
* rather than doing a complete new request regardless, if a range is partly satisfied by the cache, only request a range that fills in the gaps (e.g. if we have 0-20 and request 10-30, only request 20-30). Again, since overlapping ranges seem like an uncommon case, I didn't bother.